src/test/java/io/cdap/plugin/batch/source/MongoDBSourceConfigBuilder.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.batch.source;
+
+import io.cdap.plugin.ErrorHandling;
+import io.cdap.plugin.MongoDBConfig;
+
+/**
+ * Builder class that provides handy methods to construct {@link MongoDBBatchSource.MongoDBSourceConfig} for testing.
+ */
+public class MongoDBSourceConfigBuilder {
+
+  protected final MongoDBBatchSource.MongoDBSourceConfig config;
+
+  public static MongoDBSourceConfigBuilder builder() {
+    return new MongoDBSourceConfigBuilder(new MongoDBBatchSource.MongoDBSourceConfig());
+  }
+
+  public static MongoDBSourceConfigBuilder builder(MongoDBConfig original) {
+    return builder()
+      .setReferenceName(original.referenceName)
+      .setHost(original.host)
+      .setPort(original.port)
+      .setDatabase(original.database)
+      .setCollection(original.collection)
+      .setUser(original.user)
+      .setPassword(original.password)
+      .setConnectionArguments(original.connectionArguments);
+  }
+
+  public static MongoDBSourceConfigBuilder builder(MongoDBBatchSource.MongoDBSourceConfig original) {
+    return builder((MongoDBConfig) original)
+      .setSchema(original.schema)
+      .setInputQuery(original.inputQuery)
+      .setAuthConnectionString(original.authConnectionString)
+      .setOnError(ErrorHandling.fromDisplayName(original.onError));
+  }
+
+  private MongoDBSourceConfigBuilder(MongoDBBatchSource.MongoDBSourceConfig config) {
+    this.config = config;
+  }
+
+  public MongoDBSourceConfigBuilder setReferenceName(String referenceName) {
+    this.config.referenceName = referenceName;
+    return this;
+  }
+
+  public MongoDBSourceConfigBuilder setHost(String host) {
```

Comment:
Most of this class is duplicated in MongoDBConfigBuilder. 

It's better to parameterize MongoDBConfigBuilder:

```
public class MongoDBConfigBuilder<T extends MongoDBConfigBuilder>

  public T setHost(host) {
    this.host = host;
    return (T) this;
  }
```

then make the source builder extend it:

```
public class MongoDBSourceConfigBuilder extends MongoDBConfigBuilder<MongoDBSourceConfigBuilder> {
  ...
}
```

and only have method for the fields that are not in the base class.

Reply:
Parameterized. Thanks.